### PR TITLE
Feature/recursive members

### PR DESCRIPTION
**Pull request by TomAugspurger**
(This currently depends on #2117. I'll rebase once that's merged to v3)

This PR adds a recursive=True flag to Group.members, for recursively
listing the members of some hierarchy. The default is unchanged (just list immediate children).

**This includes a breaking change to Group.members and Group.nmembers**. Previously, these were properties, while AsyncGroup.members / nmembers were functions. I've changed the sync versions to align with the async versions (which is needed to support this keyword).
    
This is useful for [Consolidated Metadata](https://github.com/zarr-developers/zarr-python/pull/2113), which needs to recursively
inspect children. IMO, it's useful (and simple) enough to include
in the public API.

Closes https://github.com/zarr-developers/zarr-python/issues/2119

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by TomAugspurger** [comment]
https://github.com/zarr-developers/zarr-python/issues/2119 has some information on the failed hypothesis test. I *think* it's real.

**Comment by TomAugspurger** [comment]
I'm pretty stumped by the failure at https://github.com/zarr-developers/zarr-python/actions/runs/10549846136/job/29225363145?pr=2118#step:6:374

The issue comes from a `RuntimeError` in numpy when doing something like

```python
np.array([np.complex128(0.0, np.nan)]) == np.complex128(0)
```

The actual array constructed by hypothesis is a bit different though. It's more like:

```python
In [23]: np.ndarray((1,), np.dtype("complex128"), b"\x00\x00\x00\x00\x00\x00\x00\x00\xfd\x7f\xbaUFJ\xf0\x7f")[0]
Out[23]: np.complex128(nanj)
```

The dtype and repr appear to be the same, but apparently they're different internally. Not sure how yet.

Finally, something in the test runner appears to elevate the `RuntimeWarning` to an error:

```python
# inside the test run
array([False])
(Pdb) np.array([np.complex128(0.0, np.nan)]) == np.complex128(0)
array([False])
(Pdb) np.ndarray((1,), np.dtype("complex128"), b"\x00\x00\x00\x00\x00\x00\x00\x00\xfd\x7f\xbaUFJ\xf0\x7f") == np.complex128(0)
*** RuntimeWarning: invalid value encountered in equal
```

**Comment by TomAugspurger** [comment]
I'm partly giving up on that second hypothesis failure. It's possible that there's still a bug in `Buffer.all_equal` at https://github.com/zarr-developers/zarr-python/blob/681f2866343963f8fff3123f2ac8e18e2bc4b3ad/src/zarr/core/buffer.py#L429-L430 not handling `NaN` like `other` values. But for this PR, we're able to work around it by filtering out that RuntimeWarning.

This should be ready for review now.

File: tests/v3/test_properties.py [hunk 1]
```diff
@@ -32,6 +37,11 @@ def test_basic_indexing(data):
 
 
 @given(data=st.data())
+# The filter warning here is to silence an occasional warning in NDBuffer.all_equal
+# See https://github.com/zarr-developers/zarr-python/pull/2118#issuecomment-2310280899
+# Uncomment the next line to reproduce the original failure.
+# @reproduce_failure('6.111.2', b'AXicY2FgZGRAB/8/eLmF7qr/C5EDADZUBRM=')
+@pytest.mark.filterwarnings("ignore::RuntimeWarning")
```
Comment (jhamman) [inline tests/v3/test_properties.py:44]: @dcherian - would you mind looking into this?

**Review by jhamman — COMMENTED** [review]
> This includes a breaking change to Group.members and Group.nmembers. Previously, these were properties, while AsyncGroup.members / nmembers were functions. I've changed the sync versions to align with the async versions (which is needed to support this keyword).

Agreed that this is useful enough to justify this change.

File: tests/v3/test_properties.py (hunk 1 above)
Comment (TomAugspurger) [reply tests/v3/test_properties.py:44]: It's a minefield 😬 The deepest I got was `hypothesis.extra.numpy:ArrayStrategy.set_element`. Something about the sequence of operations and the data passed into the ndarray made it "weird", such that `array == np.complex128(0.0)` raised an InvalidComparision warning. I'm not sure how to interpret the bytes at https://github.com/zarr-developers/zarr-python/pull/2118#issuecomment-2310280899.

**Comment by TomAugspurger** [comment]
I was wondering about a limit-style argument too. I'll go ahead and implement that.

**Comment by TomAugspurger** [comment]
I opted to implement this as a `max_depth: int | None` keyword, which replaces `recursive=True`. By setting `max_depth=-1` or `max_depth=None` you get the same behavior as `recursive=True`. A positive integer considers just that depth. The default of `max_depth=0` is the same as the existing behavior on `main` of just considering immediate children.

**Comment by d-v-b** [comment]
thanks @TomAugspurger! what's the reasoning for specifying the same behavior with two distinct values -- `max_depth=-1` and `max_depth=None`? Personally I would be fine with XOR on these options, and it would make things marginally simpler. But maybe there's an advantage to your current approach that I'm not considering.

**Comment by TomAugspurger** [comment]
Purely for user convenience. Neither `max_depth=-1` nor `max_depth=None` sounded obviously more natural to express "no depth".

If I had to choose, I'd say `max_depth=None`. Do you have a preference?

**Comment by d-v-b** [comment]
> If I had to choose, I'd say `max_depth=None`. Do you have a preference?

I like this too!

**Review by d-v-b — APPROVED** [review]
this looks good, pre-commit is red but I don't think that failure is related to content in the PR. If you're all done here @TomAugspurger then I am happy to merge

**Comment by TomAugspurger** [comment]
Yeah, this should be all set.

The pre-commit errors are at https://github.com/zarr-developers/zarr-python/issues/2131, I think from a new version of NumPy. I'll look into them today.

**Pull request merged**